File: SimpleCADAPI/src/simplecadapi/scene/transforms.py
"""Deterministic rigid-transform primitives used by Scene validation."""

from __future__ import annotations

from typing import Mapping, Sequence


RigidTransform = Mapping[str, Sequence[float]]


def compose_rigid_transforms(
    left: RigidTransform, right: RigidTransform
) -> dict[str, list[float]]:
    """Compose parent-relative transforms using fixed multiply/add order."""

    def rotate(vector: Sequence[float]) -> list[float]:
        return [
            (left["x_axis"][index] * vector[0] + left["y_axis"][index] * vector[1])
            + left["z_axis"][index] * vector[2]
            for index in range(3)
        ]

    rotated_origin = rotate(right["origin"])
    return {
        "origin": [
            left["origin"][index] + rotated_origin[index] for index in range(3)
        ],
        "x_axis": rotate(right["x_axis"]),
        "y_axis": rotate(right["y_axis"]),
        "z_axis": rotate(right["z_axis"]),
    }


def rigid_transforms_equal(actual: RigidTransform, expected: RigidTransform) -> bool:
    """Compare derived transforms with the frozen Scene numeric profile."""

    origin_epsilon = max(
        1e-9,
        1e-12
        * max(
            1.0,
            *(abs(component) for component in actual["origin"]),
            *(abs(component) for component in expected["origin"]),
        ),
    )
    return all(
        abs(left - right) <= origin_epsilon
        for left, right in zip(actual["origin"], expected["origin"])
    ) and all(
        abs(left - right) <= 1e-12
        for key in ("x_axis", "y_axis", "z_axis")
        for left, right in zip(actual[key], expected[key])
    )


__all__ = ["compose_rigid_transforms", "rigid_transforms_equal"]

File: SimpleCADAPI/scene-contract/src/transforms.ts
/** Deterministic rigid-transform primitives used by Scene validation. */

type JsonRecord = Record<string, unknown>;

export function composeRigidTransforms(left: JsonRecord, right: JsonRecord): JsonRecord {
  const origin = left.origin as number[];
  const xAxis = left.x_axis as number[];
  const yAxis = left.y_axis as number[];
  const zAxis = left.z_axis as number[];
  const rotate = (vector: number[]): number[] => [0, 1, 2].map((index) =>
    (xAxis[index] * vector[0] + yAxis[index] * vector[1]) + zAxis[index] * vector[2]
  );
  const rotatedOrigin = rotate(right.origin as number[]);
  return {
    origin: origin.map((component, index) => component + rotatedOrigin[index]),
    x_axis: rotate(right.x_axis as number[]),
    y_axis: rotate(right.y_axis as number[]),
    z_axis: rotate(right.z_axis as number[]),
  };
}

export function rigidTransformsEqual(actual: JsonRecord, expected: JsonRecord): boolean {
  const actualOrigin = actual.origin as number[];
  const expectedOrigin = expected.origin as number[];
  const originEpsilon = Math.max(
    1e-9,
    1e-12 * Math.max(1, ...actualOrigin.map(Math.abs), ...expectedOrigin.map(Math.abs)),
  );
  return actualOrigin.every((component, index) => Math.abs(component - expectedOrigin[index]) <= originEpsilon) &&
    ["x_axis", "y_axis", "z_axis"].every((key) =>
      (actual[key] as number[]).every((component, index) =>
        Math.abs(component - (expected[key] as number[])[index]) <= 1e-12
      )
    );
}
